commons/src/main/java/com/github/phantazmnetwork/commons/component/DependencyProvider.java

```diff
@@ -0,0 +1,112 @@
+package com.github.phantazmnetwork.commons.component;
+
+import com.github.phantazmnetwork.commons.Namespaces;
+import com.github.phantazmnetwork.commons.component.annotation.ComponentDependency;
+import net.kyori.adventure.key.Key;
+import net.kyori.adventure.key.Keyed;
+import org.intellij.lang.annotations.Subst;
+import org.jetbrains.annotations.NotNull;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.function.Function;
+
+/**
+ * Represents a provider of dependencies (arbitrary data needed by a component that is not itself part of its
+ * configuration).
+ */
+public interface DependencyProvider {
+    /**
+     * The shared, always-empty DependencyProvider implementation.
+     */
+    DependencyProvider EMPTY = new DependencyProvider() {
+        @Override
+        public <TDependency> TDependency provide(@NotNull Key key) {
+            throw new IllegalArgumentException();
+        }
+
+        @Override
+        public boolean load(@NotNull Iterable<? extends Key> dependencies) {
+            return !dependencies.iterator().hasNext();
+        }
+
+        @Override
+        public boolean hasLoaded(@NotNull Key key) {
+            return false;
+        }
+    };
+
+    /**
+     * Creates a new {@link DependencyProvider} implementation that will lazily resolve dependencies using the given
+     * function during the "prepare" phase.
+     *
+     * @param dependencyFunction the function used to create dependencies
+     * @return a new DependencyProvider implementation
+     */
+    static @NotNull DependencyProvider lazy(@NotNull Function<? super Key, ?> dependencyFunction) {
+        return new LazyDependencyProvider(dependencyFunction);
+    }
+
+    /**
+     * Creates a new {@link DependencyProvider} implementation capable of providing the objects in the given array as
+     * dependencies. Each object must either implement {@link Keyed}, or supply the ComponentDependency annotation with
+     * an appropriately formatted key string. If multiple dependencies have the same key, only the first object will be
+     * used.
+     *
+     * @param objects the objects to use as dependencies
+     * @return a new DependencyProvider that can provide the given objects as dependencies
+     */
+    static @NotNull DependencyProvider ofDependencies(Object... objects) {
+        if (objects == null || objects.length == 0) {
+            return new LazyDependencyProvider(key -> null);
```

Comment:
why return null for this one but throw an exception for the empty provider

Reply:
LazyDependencyProvider will thrown an exception itself if you try to load any dependency for which the function returns null, but i can use EMPTY for this too